torch/csrc/jit/tensorexpr/loopnest.cpp

```diff
@@ -1779,6 +1779,135 @@ bool LoopNest::flatten(const std::vector<For*>& loops, For** flattened) {
   return true;
 }
 
+void LoopNest::compressBuffer(Buf* buf, Stmt* stmt) {
+  if (buf->initializer()) {
+    throw malformed_input("Can't compress buffer whose initializer is set");
+  }
+
+  // Loop iterations in NNC IR do not follow sequential semantics by default.
+  // In other words, the iterations of the loops could be executed in any
+  // random order without affecting correctness. This constraint in turn
+  // implies that there can’t be any *inter-iteration* dependences
+  // (or *loop-carried* dependences) in NNC loops. So, any NNC IR with such
+  // dependences is considered invalid.
+  //
+  // Given the constraint above, for any pair of accesses to a buffer (where
+  // at least one of the access is a write), the accesses must be
+  // loop-independent on the innermost loop containing the accesses as well as
+  // all the loops above it. So, any dimension that uses only those loop
+  // variables to access the given buffer could be optimized away.
+  //
+  // Algorithm:
+  //   * Find all the accesses to the given buf. (A)
+  //   * Find the parent common to all accesses in A. (P)
+  //   * Collect all the loops above P. (L)
+  //   * Collect all the loop variables corresponding to L. (LV)
+  //   * For every access a in A:
+  //      * For the index I in every dimension of a:
+  //          * If the variables in I are all in LV, mark this dimension
+  //            for deletion.
+  //   * For every dimension that is marked for deletion in ALL accesses in A:
+  //      * Update the buffer to set the size of that dimension to 1.
+  //      * Update all accesses in A to set the index in that dimension to 0.
+
+  auto writes = WritesToBuf::find(stmt, buf);
+  auto reads = StmtsReadingBuf::find(stmt, buf);
+
+  // All buffers must be read and written at least once.
+  // Is this a valid assumption? TODO
+  TORCH_INTERNAL_ASSERT(!writes.empty());
+  TORCH_INTERNAL_ASSERT(!reads.empty());
+
+  // Find the parent common to all the buffer accesses.
+  const Block* parent = dynamic_cast<Block*>(writes.front()->get_parent());
+  TORCH_INTERNAL_ASSERT(parent);
+  for (auto w : writes) {
+    parent = Block::getSharedParent(parent, w);
+  }
+  for (auto r : reads) {
+    parent = Block::getSharedParent(parent, r);
+  }
+  std::cout << "common parent: " << *parent << std::endl;
+
+  // Collect all the loops that are above the common parent.
+  auto loops = LoopNest::getEnclosingLoopNest(parent);
+  std::unordered_set<const Var*> loop_vars;
+  for (auto l : loops) {
+    loop_vars.insert(l->var());
+    std::cout << "enclosing loop: " << *l << std::endl;
```

Comment:
Please use `GRAPH_UPDATE/GRAPH_DEBUG` or delete (here and in other places).

Reply:
Removed.